File: rgb_led/examples/tcp_led.erl
-module(tcp_led).
-export([start/4]).

start(Port, Red, Green, Blue) ->
  {ok, LSock} = gen_tcp:listen(Port, [binary, {packet, 0}, 
                                      {active, false}]),
  do_accept(LSock, {Red, Green, Blue}).

do_accept(LSock, {Red, Green, Blue}) ->
  {ok, Sock} = gen_tcp:accept(LSock),
  {ok, Pid} = rgb_led_server:start(Red, Green, Blue),
  do_recv(Sock, Pid),
  do_accept(LSock, {Red, Green, Blue}).

do_recv(Sock, Pid) ->
  case gen_tcp:recv(Sock, 0) of
    {ok, Data} ->
      handle_data(Data, Pid),
      do_recv(Sock, Pid);
    {error, closed} ->
      ok
  end.

handle_data(Data, Pid) ->
  {Red, Green, Blue} = extract_values(Data),
  Pid ! {update, [{red, Red}, {green, Green}, {blue, Blue}]},
  Pid ! blast.

extract_values(Data) ->
  {match, [Red, Green, Blue]} = re:run(Data, value_extractor(), [{capture, ['RED', 'GREEN', 'BLUE'], list}]),
  io:format("red: ~p, green: ~p, blue: ~p\n", [Red, Green, Blue]),
  {input_to_num(Red), input_to_num(Green), input_to_num(Blue)}.

value_extractor() ->
  "R(?<RED>[0-9.]+)G(?<GREEN>[0-9.]+)B(?<BLUE>[0-9.]+).*".

input_to_num(N) ->
  case string:to_float(N) of
    {error,no_float} -> list_to_integer(N);
    {F,_Rest} -> F
  end.

File: elixir/rgb_led/lib/tcp_led.ex
defmodule TcpLed do
  alias RgbLed.Server

  def start(port, red, green, blue) do
    IO.puts "start"
    {:ok, lsock} = :gen_tcp.listen(port, [:binary, {:packet, 0}, 
                                         {:active, false}])
    do_accept(lsock, {red, green, blue})
  end

  def do_accept(lsock, {red, green, blue}) do
    IO.puts "do_accept"
    {:ok, sock} = :gen_tcp.accept(lsock)
    {:ok, pid} = Server.start([red, green, blue])
    do_recv(sock, pid)
    do_accept(lsock, {red, green, blue})
  end

  def do_recv(sock, pid) do
    IO.puts "do_recv"
    case :gen_tcp.recv(sock, 0) do
      {:ok, data} ->
        IO.puts "got #{data}"
        handle_data(data, pid)
        do_recv(sock, pid)
      {:error, :closed} ->
        :ok
    end
  end

  def handle_data(data, pid) do
    {red, green, blue} = extract_values(data)
    IO.inspect "handle data, red: #{red}"
    pid |> Server.update([red: red, green: green, blue: blue])
    pid |> Server.blast
  end

  def extract_values(data) do
    {:match, [red, green, blue]} = :re.run(data, value_extractor(), [{:capture, ['RED', 'GREEN', 'BLUE'], :list}])
    {input_to_num(red), input_to_num(green), input_to_num(blue)}
  end

  def value_extractor() do
    "R(?<RED>[0-9.]+)G(?<GREEN>[0-9.]+)B(?<BLUE>[0-9.]+).*"
  end

  def input_to_num(n) do
    case :string.to_float(n) do
      {:error,:no_float} -> list_to_integer(n)
      {f,_rest} -> f
    end
  end
end
